pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,371 @@
+    _metricName = metricNamePrefix + tableNameWithType;
+    _realTimeTableDataManager = realtimeTableDataManager;
+    _clock = Clock.systemDefaultZone();
+    // Handle negative timer values
+    if (timerThreadTickIntervalMs <= 0) {
+      throw new RuntimeException(String.format("Illegal timer timeout argument, expected > 0, got=%d for table=%s",
+          timerThreadTickIntervalMs, _tableNameWithType));
+    }
+    _enablePerPartitionMetric = enablePerPartitionMetric;
+    _enableAggregateMetric = enableAggregateMetric;
+    _timerThreadTickIntervalMs = timerThreadTickIntervalMs;
+    _timer = new Timer("IngestionDelayTimerThread" + TableNameBuilder.extractRawTableName(tableNameWithType));
+    _timer.schedule(new TimerTask() {
+        @Override
+        public void run() {
+          timeoutInactivePartitions();
+        }
+      }, INITIAL_TIMER_THREAD_DELAY_MS, _timerThreadTickIntervalMs);
+    // Install callback metric
+    if (_enableAggregateMetric) {
+      _serverMetrics.addCallbackTableGaugeIfNeeded(_metricName, ServerGauge.TABLE_INGESTION_DELAY_MS,
+          () -> getMaximumIngestionDelay());
+    }
+  }
+
+  // Constructor that uses defaults
+  public IngestionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType,
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, tableNameWithType, tableDataManager, TIMER_THREAD_TICK_INTERVAL_MS,
+        "", true, true);
+  }
+
+  // Constructor that takes a prefix to name the metric, so we can keep multiple trackers for the same table
+  public IngestionDelayTracker(ServerMetrics serverMetrics, String tableNameWithType, String metricNamePrefix,
+      RealtimeTableDataManager tableDataManager) {
+    this(serverMetrics, metricNamePrefix + tableNameWithType, tableDataManager,
+        TIMER_THREAD_TICK_INTERVAL_MS, metricNamePrefix, true, true);
+  }
+
+  /**
+   * Function that enable use to set predictable clocks for testing purposes.
+   *
+   * @param clock clock to be used by the class
+   */
+  @VisibleForTesting
+  void setClock(Clock clock) {
+    _clock = clock;
+  }
+
+  /*
+   * Called by LLRealTimeSegmentDataManagers to post delay updates to this tracker class.
+   *
+   * @param delayInMilliseconds ingestion delay being recorded.
+   * @param sampleTime sample time.
+   * @param partitionGroupId partition ID for which this delay is being recorded.
+   */
+  public void updateIngestionDelay(long delayInMilliseconds, long sampleTime, int partitionGroupId) {
+    // Store new measure and wipe old one for this partition
+    DelayMeasure previousMeasure = _partitionToDelaySampleMap.put(partitionGroupId,
+        new DelayMeasure(sampleTime, delayInMilliseconds));
+    if ((previousMeasure == null) && _enablePerPartitionMetric) {
+      // First time we start tracking a partition we should start tracking it via metric
+      _serverMetrics.addCallbackTableGaugeIfNeeded(getPerPartitionMetricName(partitionGroupId),
```

Comment:
You can add method `addPartitionCallbackGauge` in `AbstractMetrics` and making it a standard to append partitionId to the metric name. Then, you can get rid of the `getPerPartitionMetricName` method here.

Reply:
Good suggestion, addressed in ecfcf1b1cdf81f4fe3547e1eec4ac23a1df72c5c